core/cache/charmconfigwatcher.go

```diff
@@ -0,0 +1,278 @@
+	// branchChangeTopic is the pub/sub topic to which the watcher will
+	// listen for model branch change messages.
+	branchChangeTopic string
+	// branchDeleteTopic is the pub/sub topic to which the watcher will
+	// listen for model branch deletion messages.
+	branchDeleteTopic string
+
+	// hub is the pub/sub hub on which the watcher will receive messages
+	// before determining whether to notify.
+	hub *pubsub.SimpleHub
+	// res is the cache resident responsible for creating this watcher.
+	res *Resident
+}
+
+// CharmConfigWatcher watches application charm config on behalf of a unit.
+// The configuration is based on whether the unit is tracking an active model branch.
+// If keys are specified the watcher only signals a change when at least one
+// of those keys changes value. If no keys are specified,
+// any change in the config will trigger the watcher to notify.
+type CharmConfigWatcher struct {
+	*notifyWatcherBase
+
+	// initComplete is a channel that will be closed when the
+	// watcher is fully constructed and ready to handle events.
+	initComplete chan struct{}
+
+	unitName   string
+	appName    string
+	branchName string
+
+	masterSettings map[string]interface{}
+	branchDeltas   settings.ItemChanges
+	configHash     string
+}
+
+// newUnitConfigWatcher returns a new watcher for the input config keys
+// with a baseline hash of their config values from the input hash cache.
+// As per the cache requirements, hashes are only generated from sorted keys.
+func newCharmConfigWatcher(cfg charmConfigWatcherConfig) (*CharmConfigWatcher, error) {
+	w := &CharmConfigWatcher{
+		notifyWatcherBase: newNotifyWatcherBase(),
+		initComplete:      make(chan struct{}),
+		unitName:          cfg.unitName,
+		appName:           cfg.appName,
+	}
+
+	deregister := cfg.res.registerWorker(w)
+
+	multi := cfg.hub.NewMultiplexer()
+	multi.Add(cfg.appConfigChangeTopic, w.appConfigChanged)
+	multi.Add(cfg.branchChangeTopic, w.branchChanged)
+	multi.Add(cfg.branchDeleteTopic, w.branchDeleted)
+
+	w.tomb.Go(func() error {
+		<-w.tomb.Dying()
+		multi.Unsubscribe()
+		deregister()
+		return nil
+	})
+
+	if err := w.init(cfg.model); err != nil {
+		_ = w.Stop()
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// init determines baseline master configuration, branch tracking settings
+// and the configuration hash for the watcher's unit.
+// It then closes the init channel to indicate the watcher is operational.
+func (w *CharmConfigWatcher) init(model charmConfigModel) error {
+	app, err := model.Application(w.appName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	w.masterSettings = app.Config()
+
+	for _, b := range model.Branches() {
+		unitNames := b.AssignedUnits()[w.appName]
+		if len(unitNames) == 0 {
+			continue
+		}
+
+		// A unit can only track one branch at a time,
+		// so if we find one, we are done.
+		if set.NewStrings(unitNames...).Contains(w.unitName) {
+			w.branchName = b.Name()
+			w.branchDeltas = b.Config()[w.appName]
```

Comment:
Do you care if it's empty or not, I guess not. I find that `b.Config()[w.appName]` is a bit of mouthful, any chance we can split it out?

Reply:
I don't particularly mind this syntax but I agree that since we repeat this access pattern in the code below perhaps it would make sense to have a kind of `branch.ConfigForApp` helper to make this a bit easier on the eyes...

Reply:
I've added both a modularisation for the checking and a helper for the config.